src/win32_MarieAssembler.c

```diff
@@ -38,33 +38,238 @@ size_t Platform_GetFileSize(wchar_t *FileName, int *Success) {
+	"  --symboltable [FileName] ==> Outputs a file containing a symbol table for the program at [FileName], or if blank <InFileName>.sym\n"
+	"  --listing [FileName] ==> Outputs a file containing a listing for the program at [FileName], or if blank <InFileName>.lst\n";
+
+int wmain(int ArgCount, wchar_t **Args, wchar_t **Env) {
+	FILE *InFile = 0, *OutLogisim = 0, *OutHex = 0, *OutSymbolTable = 0, *OutListing = 0;
+	int GenLogisim = FALSE, GenHex = FALSE, GenSymbolTable = FALSE, GenListing = FALSE;
+	wchar_t *InFileName = 0;
+	uint64_t InFileSize = 0;
+	int Success = TRUE;
+
+	if (ArgCount == 1) {
+		printf(HelpMessage);
+		Success = FALSE;
+	}
+	
+	for (int Index = 1; Index < ArgCount;) {
+		wchar_t * Arg = Args[Index];
+		
+		if (StartsWith(Arg, L"--logisim")) {
+			if (Index + 1 >= ArgCount) { Arg = L""; }
+			else { Arg = Args[Index + 1]; }
+
+			if (OutLogisim == 0 && GenLogisim == FALSE) {
+				if (!((StartsWith(Arg, L"--")) || (Arg[0] == 0))) {
+					Index++;
+					OutLogisim = _wfopen(Arg, L"w");
+					if (OutLogisim == 0) {
+						wprintf(L"I could not open the Logisim output file \"%s\" for writing!\n", Arg);
+						Success = FALSE;
+						break;
+					}
+				}
+				else {
+					GenLogisim = TRUE;
+				}
+			}
+			else {
+				wprintf(L"Option --logisim was provided twice!\n");
+				Success = FALSE;
+				break;
+			}
+		}
+		else if (StartsWith(Arg, L"--rawhex")) {
+			if (Index + 1 >= ArgCount) { Arg = L""; }
+			else { Arg = Args[Index + 1]; }
 
-	wchar_t *StartOfFileName = PathFindFileName(Path);
-	int OutputNameLength = IndexOfFromEnd(StartOfFileName, L'.');
+			if (OutHex == 0 && GenHex == FALSE) {
+				if (!((StartsWith(Arg, L"--")) || (Arg[0] == 0))) {
+					Index++;
+					OutHex = _wfopen(Arg, L"wb");
+					if (OutHex == 0) {
+						wprintf(L"I could not open the raw hex output file \"%s\" for writing!\n", Arg);
+						Success = FALSE;
+						break;
+					}
+				}
+				else {
+					GenHex = TRUE;
+				}
+			}
+			else {
+				wprintf(L"Option --rawhex was provided twice!\n");
+				Success = FALSE;
+				break;
+			}
+		}
+		else if (StartsWith(Arg, L"--symboltable")) {
+			if (Index + 1 >= ArgCount) { Arg = L""; }
+			else { Arg = Args[Index + 1]; }
 
-	int PathLength = 0;
-	for(int Index = 0; Path[Index] != L'\0'; Index++) {
-		if (Path[Index] == L'\\' ||
-		    Path[Index] == L'/') { // @TODO I'm not totally sure if this is kosher on windows.
-			PathLength = Index + 1;
+			if (OutSymbolTable == 0 && GenSymbolTable == FALSE) {
+				if (!((StartsWith(Arg, L"--")) || (Arg[0] == 0))) {
+					Index++;
+					OutSymbolTable = _wfopen(Arg, L"w,ccs=UNICODE");
+					if (OutSymbolTable == 0) {
+						wprintf(L"I could not open the symbol table output file \"%s\" for writing!\n", Arg);
+						Success = FALSE;
+						break;
+					}
+				}
+				else {
+					GenSymbolTable = TRUE;
+				}
+			}
+			else {
+				wprintf(L"Option --symboltable was provided twice!\n");
+				Success = FALSE;
+				break;
+			}
+		}
```

Comment:
Man, I never knew how much work goes into these things... isn't there a library that we could import? I know of the existence of [Getopt](https://www.gnu.org/software/libc/manual/html_node/Example-of-Getopt.html), is there nothing similar for Windows?

Reply:
I saw a win32 port of Getopt, but it's all just straight forward parsing. Especially since we don't do options like `-ab` or `-oC:\Output`

Reply:
Personally, I'd use it just so I would minimize work and save time (plus cover any edge cases I wouldn't think of). Still a good implementation though.